File: 2962-count-subarrays-where-max-element-appears-at-least-k-times/2962-count-subarrays-where-max-element-appears-at-least-k-times.kt
class Solution {
    fun countSubarrays(nums: IntArray, k: Int): Long {
        val n = nums.size
        val max = nums.max()

        var maxCount = 0
        var s = 0
        var answer = 0L
        for (e in nums.indices) {
            if (nums[e] == max) {
                maxCount++
            }

            while (maxCount >= k && s <= e) {
                if (nums[s] == max) {
                    maxCount--
                }

                answer += n - e
                s++
            }
        }

        return answer
    }
}

File: 2962-count-subarrays-where-max-element-appears-at-least-k-times/2962-count-subarrays-where-max-element-appears-at-least-k-times.py
class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        n = len(nums)
        m = max(nums)
        r = 0
        count = 0
        answer = 0
        for l in range(n):
            while r < n and count < k:
                if nums[r] == m:
                    count += 1
                r += 1
            
            if count >= k:
                answer += n - r + 1

            if nums[l] == m:
                count -= 1

        return answer
